Mathlib/Tactic/DeriveFintype.lean

```diff
@@ -0,0 +1,315 @@
+inductive type with `n` constructors has a proxy type of the form `C₁ ⊕ (C₂ ⊕ (⋯ ⊕ Cₙ))`,
+so mapping to and from `Cᵢ` requires looking through `i` levels of `Sum` constructors.
+Ignoring time spent looking through these constructors, the construction of `Finset.univ`
+contributes linear time with respect to the cardinality of the type since the instances
+involved compute the underlying `List` for the `Finset` as `l₁ ++ (l₂ ++ (⋯ ++ lₙ))` with
+right associativity.
+
+This implementation takes some inspiration from the one by Mario Carneiro for Mathlib 3.
+A difference is that the Mathlib 3 version does not explicitly construct the total proxy type,
+and instead it opts to construct the underlying `Finset` as a disjoint union of the `Finset.univ`
+for each individual constructor's proxy type.
+-/
+
+namespace Mathlib.Deriving.Fintype
+open Lean Elab Lean.Parser.Term
+open Meta Command
+
+/-- Returns the portion of the proxy type for the constructor with arguments `xs`,
+the list of names used in the pattern, and the pattern for matching an element of this type. -/
+def mkCtorType (xs : List Expr) : TermElabM (Expr × List Name × TSyntax `term) :=
+  match xs with
+  | [] => return (mkConst ``Unit, [], ← `(term| ()))
+  | [x] => do
+    let a ← mkFreshUserName `a
+    return (← inferType x, [a], mkIdent a)
+  | x :: xs => do
+    let (ty, names, patt) ← mkCtorType xs
+    let ty ← mkAppM ``PSigma #[← mkLambdaFVars #[x] ty]
+    let a ← mkFreshUserName `a
+    return (ty, a :: names, ← `(term| ⟨$(mkIdent a), $patt⟩))
+
+/-- Navigates into the sum type that we create in `mkCType` for the given constructor index. -/
+def wrapSumAccess (cidx nctors : Nat) (spatt : TSyntax `term) : TermElabM (TSyntax `term) :=
+  match cidx with
+  | 0 =>
+    if nctors = 1 then
+      return spatt
+    else
+      `(term| Sum.inl $spatt)
+  | cidx' + 1 => do
+    let spatt ← wrapSumAccess cidx' (nctors - 1) spatt
+    `(term| Sum.inr $spatt)
+
+/-- Create a `Sum` of types, mildly optimized to not have a trailing `Empty`.
+Returns also the necessary tactic for the `left_inv` proof in the `Equiv` we construct. -/
+def mkCType (ctypes : List Expr) : TermElabM (Expr × TSyntax `tactic) :=
+  match ctypes with
+  | [] => return (mkConst ``Empty, ← `(tactic| cases x))
+  | [x] => return (x, ← `(tactic| rfl))
+  | x :: xs => do
+    let (ty, pf) ← mkCType xs
+    let pf ← `(tactic| cases x with | inl _ => rfl | inr x => $pf:tactic)
+    return (← mkAppM ``Sum #[x, ty], pf)
+
+structure EquivData where
+  /-- Name of the declaration for a type that is `Equiv` to the given type. -/
+  proxyName : Name
+  /-- A term that elaborates to something of type `proxyType ≃ type`.
+  Needs to be elaborated with this expected type. -/
+  equiv : TSyntax `term
+
+/--
+Generates a proxy type for the inductive type (defining it at `fintypeProxy` in the
+inductive type's namespace) and returns an equivalence to it.
+-/
```

Comment:
Can you extend this with an example of what the generated proxy type is?

Reply:
I've added an example on `proxy_equiv%`.